backend/contest/room.go

```diff
@@ -169,6 +170,18 @@ func (r *roomImpl) Join(name string) Player {
 	return player
 }
 
+func (r *roomImpl) Leave(playerKey string) Player {
+	player := r.players[playerKey]
+	r.notifyPlayers(
+		func(p Player) {
+			p.NotifyPlayerLeft(player.Name, player.Key)
+		},
+	)
+	delete(r.players, playerKey)
+	// TODO cant return *Player why?
```

Comment:
`r.players` is a map of strings to _pointers_ to players.

You declare the return type `Player`.

Thus, you have a type mismatch "pointer to player" vs "Player". You can either:
1. have return type `Player`, then you have to `return *player` (the '*' dereferences the pointer)
2. have return type `*Player`, then you can directly `return player`

The `r.players` map uses pointers to make the interface embedding with the `Notifier` work (line 367) – one of Go's many quirks.

I favor the first option because it prevents the player from being changed outside the package (`Leave` is an exported method). Unfortunately, the `FindPlayer` method violates this principle, because one of the method's caller needs the `Notifier` embedding and thus needs a pointer :(.

Reply:
Thanks for the detailed explanation.

I also prefer the first solution.